ceph/cephfs/cephfs_provisioner/cephfs_provisioner.py

```diff
@@ -199,13 +220,69 @@ def cap_update(orig, want, unwanted):
         assert caps[0]['entity'] == client_entity
         return caps[0]
 
+    def _create_volume(self, volume_path, size=None, data_isolated=False, namespace_enforced=True):
+        """
+        Linux kernel cephfs only support Rados Pool Namespace after 4.8:
+        https://github.com/torvalds/linux/commit/779fe0fb8e1883d5c479ac6bd85fbd237deed1f7.
+        This methid is modified from
+        https://github.com/ceph/ceph/blob/v13.0.0/src/pybind/ceph_volume_client.py#L603
+        and make namespace enforcement optional.
+        Set up metadata, pools and auth for a volume.
+        This function is idempotent.  It is safe to call this again
+        for an already-created volume, even if it is in use.
+        :param volume_path: VolumePath instance
+        :param size: In bytes, or None for no size limit
+        :param data_isolated: If true, create a separate OSD pool for this volume
+        :return:
+        """
+        path = self.volume_client._get_path(volume_path)
+        log.info("create_volume: {0}".format(path))
+
+        self.volume_client._mkdir_p(path)
+
+        if size is not None:
+            self.volume_client.fs.setxattr(path, 'ceph.quota.max_bytes', size.__str__(), 0)
+
+        # data_isolated means create a separate pool for this volume
+        if data_isolated:
+            pool_name = "{0}{1}".format(self.volume_client.POOL_PREFIX, volume_path.volume_id)
+            log.info("create_volume: {0}, create pool {1} as data_isolated =True.".format(volume_path, pool_name))
+            pool_id = self.volume_client._create_volume_pool(pool_name)
```

Comment:
Seems it's not suitable for each volume create one Ceph pool as it would increase the pgs on OSD. For performance recommmand, each Ceph OSD pgs should less than one number(ex. 300).

Reply:
This func supports to isolate data (it's just copied from https://github.com/ceph/ceph/blob/v13.0.0/src/pybind/ceph_volume_client.py#L603 with some modifications for our needs), in this script `data_isolated` is always `False`.